Introduce --parser option for a single benchmark

hkdobrev opened the pull request:
- Installing the parsers is optional. It is now in the `require-dev` directive.
- Checking the requirements for a driver would not throw an exception.
  It is now OK to have a missing parser class.
  Also the `checkRequirements()` method was never used until now.
- Only valid drivers (fulfilling the requirements) enter the benchmark.
- Introducing a new `--parser` option to limit the benchmark to a single parser.

The default behavior is left intact.

This allows the benchmarking tool to be run not only on its own, but also
as part of the CI cycle of the parsers themselves.

For example the Ciconia parser could have something like this in its Travis file:

```
after_success:
  - composer require kzykhys/markbench:*
  - vendor/bin/markbench benchmark --parser=Ciconia --profile=blank
  - vendor/bin/markbench benchmark --parser=Ciconia --profile=default
  - vendor/bin/markbench benchmark --parser=CiconiaGfm --profile=github-sample
```

---

Resolve #3.

hkdobrev committed:
Introduce --parser option for a single benchmark

- Installing the parsers is optional. It is now in the `require-dev` directive.
- Checking the requirements for a driver would not throw an exception.
  It is now OK to have a missing parser class.
  Also the `checkRequirements()` method was never used until now.
- Only valid drivers (fulfilling the requirements) enter the benchmark.
- Introducing a new `--parser` option to limit the benchmark to a single parser.
  Use the name of the parser or limit it to a dialect with colon-separator like so:
      bin/markbench benchmark --parser=kzykhys/ciconia:gfm

The default behavior is left intact.

This allows the benchmarking tool to be run not only on its own, but also
as part of the CI cycle of the parsers themselves.

For example the Ciconia parser could have something like this in its Travis file:

    after_success:
      - composer require kzykhys/markbench:*
      - vendor/bin/markbench benchmark --parser=kzykhys/ciconia --profile=blank
      - vendor/bin/markbench benchmark --parser=kzykhys/ciconia --profile=default
      - vendor/bin/markbench benchmark --parser=kzykhys/ciconia --profile=github-sample

hkdobrev commented on composer.json:
```diff
@@ -20,10 +20,12 @@
         "kzykhys/ciconia": "*",
         "michelf/php-markdown": "*",
         "erusev/parsedown": "*"
     },
```
I've moved the parser packages from `require` to `require-dev` so they would be installed when running the benchmark of all of them in the Markbench CI, but they would not be installed when just installing Markbench as a part of one of the parsers.

hkdobrev commented on src/Markbench/Console/Command/BenchmarkCommand.php:
```diff
@@ -150,9 +156,16 @@ protected function registerDrivers(Runner $runner)
-            $fqcn      = 'Markbench\\Driver\\' . $className;
 
-            $runner->addDriver(new $fqcn());
+            if ($parserName === static::ALL_PARSERS OR strpos(strtolower($className), $parserName) === 0)
```
As you can see, this is counting on some naming convention of the parsers. Perhaps a bigger structural change would be needed to handle the check better. My point was to allow easier use on the command line. For example:

```
bin/markbench --parser=ciconia
```

would match both `Ciconia` and `CiconiaGfm` drivers.

hkdobrev commented on src/Markbench/Console/Command/BenchmarkCommand.php:
```diff
@@ -32,7 +34,8 @@ protected function configure()
             ->setName('benchmark')
-            ->setDescription('Run a benchmark with selected profile')
+            ->setDescription('Run a benchmark with selected parser and profile')
+            ->addOption('parser', '', InputOption::VALUE_OPTIONAL, 'Name of a parser. Available: Ciconia, Parsedown, PHPMarkdown', static::ALL_PARSERS)
```
I'm kind of new to this `Symfony\Component\Console` thing so all suggestions about the default, the shorthand flag and the descriptions are more than welcome.

kzykhys replied:
We should use InputOption::VALUE_REQUIRED and remove default value

hkdobrev replied:
@kzykhys I wanted to achieve some kind of backwards compatibility and easier run of all benchmarks. So you'd be able to skip the `--parser` argument and then all parsers would be benchmarked.

kzykhys replied:
@hkdobrev Please see http://symfony.com/doc/current/components/console/introduction.html#using-command-options

We can skip `--parser` even if InputOption::VALUE_REQUIRED is set.

hkdobrev replied:
@kzykhys Thanks!

hkdobrev commented on src/Markbench/Console/Command/BenchmarkCommand.php:
```diff
@@ -43,19 +46,19 @@ protected function configure()
         $profile  = $input->getOption('profile');
 
-        if (!isset($profiles)) {
+        if (!isset($profiles[$profile])) {
```
I guess this line here fixed a small mistake from before, but I don't think it was worth another PR since it was not a problem until now.

kzykhys commented:
Good idea! Thanks!

How about using the package name instead of the class name for `--parser` option?

```
bin/benchmark benchmark --parser=erusev/parsedown
bin/benchmark benchmark --parser=kzykhys/ciconia
bin/benchmark benchmark --parser=kzykhys/ciconia:gfm
bin/benchmark benchmark --parser=michelf/php-markdown:extra
```

``` php
$driver = new $fqcn();
//$parserName = $this->getOption('parser'); // null if omitted

if (!$parserName || in_array($parser, [$driver->getName(), $driver->getName().':'.$driver->getDialect()])) {
    $runner->addDriver($driver);
}
```

kzykhys commented on src/Markbench/Console/Command/BenchmarkCommand.php:
```diff
@@ -43,19 +46,19 @@ protected function configure()
         foreach ($runner->getDrivers() as $driver) {
-            if (!isset($this->versions[$driver->getName()])) {
+            if (!isset($this->versions[$driver->getName()]) AND $driver->checkRequirements()) {
                 $this->versions[$driver->getName()] = $this->getVersion($driver->getName());
```
`&&` instead of `AND`
http://www.php.net/manual/en/language.operators.logical.php

hkdobrev commented:
@kzykhys I think I've addressed your remarks. I've removed one of the `checkRequirements()` check and added a default empty array value to `Runner::$drivers`.

kzykhys commented:
Looks great, thanks a lot!

```
hayashi$ php bin/markbench benchmark --profile=blank
+----------------------+---------+---------+---------------+---------+--------------+
| package              | version | dialect | duration (MS) | MEM (B) | PEAK MEM (B) |
+----------------------+---------+---------+---------------+---------+--------------+
| erusev/parsedown     | 0.9.0   |         | 14            | 6029312 | 6029312      |
| michelf/php-markdown | 1.4.0   |         | 219           | 6029312 | 6029312      |
| michelf/php-markdown | 1.4.0   | extra   | 253           | 6029312 | 6029312      |
| kzykhys/ciconia      | v1.0.3  |         | 847           | 6553600 | 6553600      |
| kzykhys/ciconia      | v1.0.3  | gfm     | 944           | 6553600 | 6553600      |
+----------------------+---------+---------+---------------+---------+--------------+

hayashi$ php bin/markbench benchmark --profile=blank --parser=kzykhys/ciconia
+-----------------+---------+---------+---------------+---------+--------------+
| package         | version | dialect | duration (MS) | MEM (B) | PEAK MEM (B) |
+-----------------+---------+---------+---------------+---------+--------------+
| kzykhys/ciconia | v1.0.3  |         | 835           | 5505024 | 5767168      |
| kzykhys/ciconia | v1.0.3  | gfm     | 908           | 5767168 | 5767168      |
+-----------------+---------+---------+---------------+---------+--------------+

hayashi$ php bin/markbench benchmark --profile=blank --parser=kzykhys/ciconia:gfm
+-----------------+---------+---------+---------------+---------+--------------+
| package         | version | dialect | duration (MS) | MEM (B) | PEAK MEM (B) |
+-----------------+---------+---------+---------------+---------+--------------+
| kzykhys/ciconia | v1.0.3  | gfm     | 647           | 5767168 | 5767168      |
+-----------------+---------+---------+---------------+---------+--------------+
```
